engine/execution/ingestion/engine.go

```diff
@@ -66,6 +74,8 @@ type Engine struct {
 	syncFast               bool                // sync fast allows execution node to skip fetching collection during state syncing, and rely on state syncing to catch up
 	checkAuthorizedAtBlock func(blockID flow.Identifier) (bool, error)
 	pauseExecution         bool
+	workQueueLimitCh       chan struct{} // a buffered channel of struct{} to guard max number of blocks currently being processed
+	workQueueLimitSize     uint32
```

Comment:
nit: you may consider factoring this out into something more reusable like https://github.com/remeh/sizedwaitgroup (in case you feel bounded concurrency can be a useful pattern in the codebase.)

Reply:
Nice idea. I will explore in that direction